src/models/responses/mod.rs

```diff
@@ -0,0 +1,91 @@
+use crate::models::streams::ContentLength;
+use rocket_contrib::json::Json;
+use rocket::http::{ContentType, Status};
+use rocket::response::{Responder};
+use rocket::Request;
+use serde::Serialize;
+use std::io::Read;
+
+/// A response for a file stream
+pub struct FileResponse<T: Read + 'static> {
+    file_data: Option<T>,
+    status: Status,
+    content_type: ContentType,
+    content_length: ContentLength,
+}
+
+impl<T: Read + 'static> FileResponse<T> {
+	/// Creates a new file response from the provided optional T, status,
+	/// ContentType, and ContentLength
+    pub fn new(
+        read: Option<T>,
+        status: Status,
+        content_type: ContentType,
+        content_length: ContentLength
+    ) -> Self {
+        FileResponse {
+            file_data: read,
+            status: status,
+            content_length,
+            content_type,
+        }
+    }
+
+	/// Creates a new FileResponse with an Ok Status from the provided T,
+	/// ContentType, and ContentLength
+    pub fn ok(cur: T, content_type: ContentType, content_length: ContentLength) -> Self {
+        FileResponse::new(Some(cur), Status::Ok, content_type, content_length)
+    }
+
+	/// Creates a new FileResponse with a NotFound Status.
+    pub fn not_found() -> Self {
+        FileResponse::new(None, Status::BadRequest, ContentType::default(), ContentLength(0))
+    }
+}
+
+impl<'r, T: Read> Responder<'r> for FileResponse<T> {
+    fn respond_to(self, _: &Request) -> rocket::response::Result<'r> {
+        match self.file_data {
+            Some(file_data) => {
+                let reader = file_data;
+
+                rocket::response::Response::build()
+                    .streamed_body(reader)
+                    .status(self.status)
+                    .header(self.content_type)
+                    .header(self.content_length)
+                    .ok()
+            },
+            None => {
+                rocket::response::Response::build()
+                    .status(self.status)
+                    .header(self.content_type)
+                    .header(self.content_length)
+                    .ok()
+            }
+        }
+
+
+    }
+}
+
+/// A general API response that wraps a JSON response and status code.
+pub struct ApiResponse<T> {
+    pub json: Json<Option<T>>,
```

Comment:
This can be simplified by making this field just `pub json: Option<T>`, then applying the constraint `T: Serialize`.

Then in your `Responder` implementation, you can use `Json(self.json).respond_to(req).unwrap()` to create your JSON response.

Your usages of `ApiResponse<T>` will then no longer need the `Json` struct:

```rust
ApiResponse {
    json: Some("Hello, World!"),
    status: Status::Ok,
}
```